packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,436 @@
+         : new AssetImage(placeholder, bundle: bundle),
+       image = new NetworkImage(image, scale: scale),
+       assert(fadeOutDuration != null),
+       assert(fadeOutCurve != null),
+       assert(fadeInDuration != null),
+       assert(fadeInCurve != null),
+       super(key: key);
+
+  /// Image displayed while the target [image] is loading.
+  final ImageProvider placeholder;
+
+  /// The target image that is displayed.
+  final ImageProvider image;
+
+  /// The duration of the fade-out animation for the [placeholder].
+  final Duration fadeOutDuration;
+
+  /// The curve of the fade-out animation for the [placeholder].
+  final Curve fadeOutCurve;
+
+  /// The duration of the fade-in animation for the [image].
+  final Duration fadeInDuration;
+
+  /// The curve of the fade-in animation for the [image].
+  final Curve fadeInCurve;
+
+  /// If non-null, require the image to have this width.
+  ///
+  /// If null, the image will pick a size that best preserves its intrinsic
+  /// aspect ratio.
+  final double width;
+
+  /// If non-null, require the image to have this height.
+  ///
+  /// If null, the image will pick a size that best preserves its intrinsic
+  /// aspect ratio.
+  final double height;
+
+  /// How to inscribe the image into the space allocated during layout.
+  ///
+  /// The default varies based on the other fields. See the discussion at
+  /// [paintImage].
+  final BoxFit fit;
+
+  /// How to align the image within its bounds.
+  ///
+  /// An alignment of (0.0, 0.0) aligns the image to the top-left corner of its
+  /// layout bounds.  An alignment of (1.0, 0.5) aligns the image to the middle
+  /// of the right edge of its layout bounds.
+  final FractionalOffset alignment;
+
+  /// How to paint any portions of the layout bounds not covered by the image.
+  final ImageRepeat repeat;
+
+  /// The center slice for a nine-patch image.
+  ///
+  /// The region of the image inside the center slice will be stretched both
+  /// horizontally and vertically to fit the image into its destination. The
+  /// region of the image above and below the center slice will be stretched
+  /// only horizontally and the region of the image to the left and right of
+  /// the center slice will be stretched only vertically.
+  final Rect centerSlice;
+
+  @override
+  State<StatefulWidget> createState() => new _FadeInImageState();
+}
+
+
+/// The phases a [FadeInImage] goes through.
+@visibleForTesting
+enum FadeInImagePhase {
+  /// The initial state.
+  ///
+  /// We do not yet know whether the target image is ready and therefore no
+  /// animation is necessary, or whether we need to use the placeholder and
+  /// wait for the image to load.
+  start,
+
+  /// Waiting for the target image to load.
+  waiting,
+
+  /// Fading out previous image.
+  fadeOut,
+
+  /// Fading in new image.
+  fadeIn,
+
+  /// Fade-in complete.
+  completed,
+}
+
+typedef _ImageProviderResolverListener = void Function();
```

Comment:
let's stick to the old style for now until we're ready to flip the entire codebase over to the new style